### DISCUSSION TITLE
WordPress hooks intelisense

### DISCUSSION THREAD

**Discussion by bart-jaskulski** [phpactor/phpactor#2365] [Ideas]
WordPress, as much as condemned, is a large CMS with a vast amount of plugins and themes written by developers. At the moment, PhpStorm has a very good (IMO) support for developing WordPress based projects. Similarly, VS Code has at least one plugin, which provides intellisense for the most important feature in this CMS: **hooks** (i.e. event subscribing and dispatching).

The issue in Phpactor about WordPress integration was raised 5 yr ago (#595), but it mainly focused about indexing WordPress codebase for correct completion, etc. Actually, this can be easily solved for some time with including WordPress in project's indexed paths and is very similar to what JetBrains have implemented.

The truly influential step would be to index WordPress hooks, to provide such actions as going to definition/declaration, displaying the documentation and completing the hook name. I recon, this wouldn't be that difficult if not for the shape of the hook, which is simply a string literal used as first parameter in a function call (e.g. `add_action( 'my_hook', 'my_callback' )` ).

As far as I know, introspecting on function call parameters isn't something what can be done easily at the current state of Phpactor. An only viable approach to that would be to index such hooks, in the same manner classes, methods, and functions are indexed, with declaration location and its references.

My knowledge of Phpactor isn't enough at the moment to came up with a possible implementation (regarding the proposal above actually make sense). I'd like to ask the core maintainers for guidance, how to write an indexer which is a part of extension. I've seen indexer only in a core Phpactor functionalities, and I'm not sure how to add any optional one. Anyway, with a little help I'd like to convert that to an issue and start to work on the WordPress extension.

**Comment by dantleech** [comment]
So, you want to:

- Auto-complete the hook _name_ and ideally show documentation for it?

Then:

- Where and how are hooks defined?

The Symfony extension provides completion for [service names](https://github.com/phpactor/phpactor/blob/d97e177ee7aa5800cbbf51f4113dba6a4c77b60d/lib/Extension/Symfony/Completor/SymfonyContainerCompletor.php#L21) for calls to the service locator.

**Reply by bart-jaskulski** [reply]
> The Symfony extension provides completion for [service names](https://github.com/phpactor/phpactor/blob/d97e177ee7aa5800cbbf51f4113dba6a4c77b60d/lib/Extension/Symfony/Completor/SymfonyContainerCompletor.php#L21) for calls to the service locator.

While Symfony completion is good as a starter for that feature, it is revolving around XML inspector. With WordPress it gets complicated as we (assuming no index beforehand) should scan all `do_action` references to extract the first argument from each call and provide it to the completion engine. And there actually can be loads of it: WordPress itself has 1035 actions and 2286 filters, and the most popular plugin (WooCommerce) has 1173 actions and 2208 filters. Moreover, when I first approached this issue with mundane scanning over each reference in mind, I wasn't able to extract referenced function call arguments (it might be due to my insufficient knowledge of Phpactor).

The final goal would be to support the following LSP methods:

- *textDocument/completion* - auto-complete the hook identifier
- *textDocument/hover* - show documentation based on phpdoc for the hook (ref: [WordPress core hook](https://github.com/WordPress/wordpress-develop/blob/362624176cba41a2dda57c3e89031aa6c3e4decf/src/wp-settings.php#L630-L641))
- *textDocument/definition* - navigate to hook definition, i.e. `do_action` function call with specified hook name
- *textDocument/references* - when at hook definition (`do_action`), populate references to attached hooks (`add_action` calls) with the same identifier

## When and how the hooks are defined?

Similarly to regular event dispatcher, a hook is a pair of function calls with the same string identifier. Hook call (dispatching) may be placed anywhere, the user wants anyone else to be able to attach some specific piece of code. Hook callback has to be registered anywhere **before** actual dispatch runtime.

```php
// Register a callback (second parameter), associated with event identified by name "init"
add_action('init', function ($param, $param) {
  // do some extra work here...
});

// "Dispatch" init event somewhere in the code.
// All hooks registered before this call will be executed, i.e. our callback above will do its work
do_action('init', 'some argument', new ClassArgument() /* etc... */ );
```

Hooks named *actions* are there to provide some side effects, like custom post types registration, etc.

There's also a variant called *filter* (`apply_filters`, `add_filter`), which purpose is to modify the first hook argument (second argument of function call). For the purpose of features above, the differences between actions and filters doesn't really matter, as those still are hooks (actually, in the WordPress implementation *action* **is** a filter which doesn't care about the returned value).

```php
function helloWorld() {
  return apply_filters(
    'hello_world', // our "event" identifier
    'Hello World', // hook argument to be filtered
    'en' // you can provide additional arguments
  );
}

add_filter('hello_world', function ( $string ) {
  return 'Ola Mundo';
});

helloWorld(); // returns 'Ola Mundo'
```

## Current implementations

A good reference when it comes to functionality is [PhpStorm built-in support for hooks](https://www.jetbrains.com/help/phpstorm/wordpress-aware-coding-assistance.html#hooks-support). An important part about it is at the beginning of the section:

> PhpStorm indexes all hooks declared in the WordPress core and in the included plugins. This makes the basis for coding assistance when working with hooks.

It's crucial, as there's also a VS Code extension, which relies on static hook data: https://github.com/wp-hooks/vscode-wordpress-hooks. While it may be an easier implementation, it's very limited, because any plugin and theme developer may define the own hooks, and those also should be indexed and provided for LSP actions. The mentioned extension simply parses all WordPress core hooks into JSON and builds its support on top of [those files](https://github.com/wp-hooks/wordpress-core-hooks/blob/develop/hooks/actions.json).

### Further references

- [WordPress Hooks Handbook](https://developer.wordpress.org/plugins/hooks/)

**Reply by dantleech** [reply]
so currently there is no way to extend the indexer:

https://github.com/phpactor/phpactor/blob/d97e177ee7aa5800cbbf51f4113dba6a4c77b60d/lib/Indexer/Adapter/Tolerant/TolerantIndexBuilder.php#L45

that's an easy thing to fix. what I'm not sure about is how far you can get with a custom record. In theory you can just create a Record class as with the other indexers and call `$index->write($record)`, but the design is far from perfect and I can't say off my head how far you'll get after that.
